src/zarr/storage/_common.py

```
@@ -54,59 +62,87 @@ def __init__(self, store: Store, path: str = "") -> None:
+
         Raises
         ------
         FileExistsError
             If the mode is 'w-' and the store path already exists.
         ValueError
             If the mode is not "r" and the store is read-only, or
-            if the mode is "r" and the store is not read-only.
         """
 
-        await store._ensure_open()
-        self = cls(store, path)
-
         # fastpath if mode is None
         if mode is None:
-            return self
+            return await cls._create_open_instance(store, path)
 
-        if store.read_only and mode != "r":
-            raise ValueError(f"Store is read-only but mode is '{mode}'")
-        if not store.read_only and mode == "r":
-            raise ValueError(f"Store is not read-only but mode is '{mode}'")
+        if mode not in ANY_ACCESS_MODE:
+            raise ValueError(f"Invalid mode: {mode}, expected one of {ANY_ACCESS_MODE}")
 
+        if store.read_only:
+            # Don't allow write operations on a read-only store
+            if mode != "r":
+                raise ValueError(
+                    f"Store is read-only but mode is {mode!r}. Create a writable store or use 'r' mode."
+                )
+            self = await cls._create_open_instance(store, path)
+        elif mode == "r":
+            # Create read-only copy for read mode on writable store
+            try:
+                warnings.warn(
+                    "Store is not read-only but mode is 'r'. Creating a read-only copy. "
+                    "This behavior may change in the future with a more granular permissions model.",
+                    UserWarning,
+                    stacklevel=1,
+                )
+                self = await cls._create_open_instance(store.with_read_only(True), path)
+            except NotImplementedError as e:
+                raise ValueError(
+                    "Store is not read-only but mode is 'r'. Unable to create a read-only copy of the store."
```

_User when this error is thrown:_ "But why was it unable to create a read-only copy?"

The fact it's because the store doesn't implement something specific is totally swallowed by this error message, with no further direction given. I think the re-raised error message would be more helpful if it ended with `because store class {type(store)} does not implement method ...`.

---

I'll add direction to first create a read-only store or use a store than implements `with_read_only`. I don't agree with the value of duplication in the tracebace and think users to expect to read tracebacks e.g.,:
```bash
Exception has occurred: ValueError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
Store is not read-only but mode is 'r'. Unable to create a read-only copy of the store.
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/storage/_common.py", line 126, in open
    # writable store and writable mode
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/abc/store.py", line 103, in with_read_only
        f"with_read_only is not implemented for the {type(self)} store type."
    
    )
NotImplementedError: with_read_only is not implemented for the <class 'zarr.storage._zip.ZipStore'> store type.

The above exception was the direct cause of the following exception:

  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/storage/_common.py", line 128, in open
    
    
    # Handle mode-specific operations
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/storage/_common.py", line 373, in make_store_path
    return result
    
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/api/asynchronous.py", line 1262, in open_array
    store_path = await make_store_path(store, path=path, mode=mode, storage_options=storage_options)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/core/sync.py", line 119, in _runner
    return await coro
           ^^^^^^^^^^
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/core/sync.py", line 163, in sync
    raise return_result
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/src/zarr/api/synchronous.py", line 1297, in open_array
            async_api.open_array(
            ^^^^^^^^^^^^^^^^^^^^^
    ...<7 lines>...
                storage_options=storage_options,
    
  File "/Users/max/Documents/Code/zarr-developers/zarr-python/.vscode/test-zip-mode.py", line 5, in <module>
    arr = zarr.open_array(store=store, path="a", zarr_format=2, mode="r")
  File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/runpy.py", line 88, in _run_code
    exec(code, run_globals)
  File "/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13/lib/python3.13/runpy.py", line 198, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
ValueError: Store is not read-only but mode is 'r'. Unable to create a read-only copy of the store.
```

---

You're right - I missed the fact that raising from the `NotImplementedError` wouldn't actually completely swallow it, so the user would get to see the `NotImplementedError` message and hence why it actually failed.